### PR TITLE
blockstorage: XOR blocksdir *.dat files

### PR DESCRIPTION
**Pull request by maflcko**
Currently the *.dat files in the blocksdir store the data received from remote peers as-is. This may be problematic when a program other than Bitcoin Core tries to interpret them by accident. For example, an anti-virus program or other program may scan them and move them into quarantine, or delete them, or corrupt them. This may cause Bitcoin Core to fail a reorg, or fail to reply to block requests (via P2P, RPC, REST, ...).

Fix this, similar to https://github.com/bitcoin/bitcoin/pull/6650, by rolling a random XOR pattern over the dat files when writing or reading them.

Obviously this can only protect against programs that accidentally and unintentionally are trying to mess with the dat files. Any program that intentionally wants to mess with the dat files can still trivially do so.

The XOR pattern is only applied when the blocksdir is freshly created, and there is an option to disable it (on creation), so that people can disable it, if needed.

### CODE REVIEW

**Comment by jamesob** [comment]
Concept ACK. Certainly worth doing.

**Comment by luke-jr** [comment]
AFAIK there is software which reads these files intentionally. IMO they can be expected to adapt, but having the option to disable it seems like a good idea in case they aren't ready right away.

**Comment by dergoegge** [comment]
Concept ACK

**Comment by darosior** [comment]
Concept ACK.

**Comment by willcl-ark** [comment]
Concept ACK.

I've not reviewed the code in detail yet, but this is a nice followup to #28060.

It seems to work well so far. I wrote a quick-n-dirty [python script](https://gist.github.com/willcl-ark/d1dd7d80d4581671aa38157960a87ba7) to manually xor my data dir using the same 8 byte key from this pull, and Bitcoin Core at commit fa3c1d230a didn't have any problems reading the xored blocks or key afterwards.

**Comment by kristapsk** [comment]
Concept ACK

**Review by pablomartin4btc — COMMENTED** [review]
tACK fa9f3f70011842e2c4358d15ab4639734d976f5c

<details>
<summary>Played around with the <code>xor.key</code> file, if I changed it (use case: "<code>The created XOR-key will be zeros for an existing blocksdir</code>"), I got the <code>Corrupted block database detected</code> error, which can be easily resolved just deleting the file and running the node back up again will re-create the <code>xor.key</code> file correctly.</summary>

```
2023-10-26T18:07:06Z init message: Verifying blocks…
2023-10-26T18:07:06Z Verifying last 6 blocks at level 3
2023-10-26T18:07:06Z Verification progress: 0%
2023-10-26T18:07:06Z ERROR: ReadBlockFromDisk: Deserialize or I/O error - ReadCompactSize(): size too large: iostream error at FlatFilePos(nFile=5, nPos=62189381)
2023-10-26T18:07:06Z Verification error: ReadBlockFromDisk failed at 160000, hash=0000003ca3c99aff040f2563c2ad8f8ec88bd0fd6b8f0895cfaf1ef90353a62c
2023-10-26T18:07:06Z : Corrupted block database detected.
Please restart with -reindex or -reindex-chainstate to recover.
: Corrupted block database detected.
Please restart with -reindex or -reindex-chainstate to recover.
2023-10-26T18:07:06Z Aborted block database rebuild. Exiting.
2023-10-26T18:07:06Z Shutdown: In progress...
2023-10-26T18:07:06Z scheduler thread exit
2023-10-26T18:07:06Z Flushed fee estimates to fee_estimates.dat.
2023-10-26T18:07:06Z Shutdown: done
```

`2023-10-26T18:12:03Z Using xor key for *.dat files: '0000000000000000'`

</details>

Now, considering the possibility of the XOR key file getting corrupted (for example, accidentally opening it with an app that offers to 'fix' a corruption) or being deleted, especially on a fresh run when a random XOR key is generated (which I've also tested), wouldn't it be better to save the key in the .conf or settings?

Before having that thought, I was considering proposing to give the option to the user to specify the filename, but that would add complexity, and I'm not sure if it's worth it.

**Review by pablomartin4btc — COMMENTED** [review]
re ACK fa6b180beb20d64962c2b5440eeacb31ed889a81

**Comment by lrs955** [comment]
why not save the xor key inside the leveldb block database?

> > wouldn't it be better to save the key in the .conf or settings?
> 
> The key is not a configuration or setting that is supposed to be modified by a user or that can be changed at all. Once it is persisted to storage, it is set in stone for all (future) dat files.

**Comment by lrs955** [comment]
> > why not save the xor key inside the leveldb block database?
> 
> Just seems more complicated with no benefit?

I understand.
Maybe im wrong but isn't also the xor pattern of the chainstate saved in the database itself? Different leveldb database but still in a key a database.

**Comment by willcl-ark** [comment]
While you're re-touching, I think it could be worth adding an entry to `doc/files.md` too, especially as losing/not backing-up this file results in requiring a full reindex?

**Review by willcl-ark — APPROVED** [review]
ACK faaf3cf535999c2c9a84337414c8b35435384862

Nice change with the same rationale as the previous mempool XOR changes (preventing anti-virus falsely flagging/modifying block files).

One non-blocking comment left as a result of manual testing.

File: src/node/blockstorage.cpp [hunk 1]
```diff
@@ -1156,9 +1157,41 @@ FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, cons
...
+    }
+    // If the user disabled the key, it must be zero.
+    if (!opts.xor_key && xor_key != decltype(xor_key){}) {
+        throw std::runtime_error{
+            strprintf("The blocksdir XOR-key can not be disabled when a random key was already stored! "
+                      "Stored key: '%s', stored path: '%s'.",
+                      HexStr(xor_key), fs::PathToString(xor_key_path)),
+        };
+    }
+    LogInfo("Using xor key for *.dat files: '%s'\n", HexStr(xor_key));
```
Comment (TheCharlatan) [inline src/node/blockstorage.cpp:1188]: Nit: 
This prints as:
```
Using xor key for *.dat files: '0000000000000000'`
```
But the log lines informing the user on data obfuscation in the other DBs use:
```
Using obfuscation key for /home/drgrid/.bitcoin/blocks/index: 0000000000000000
Using obfuscation key for /home/drgrid/.bitcoin/chainstate: a3b154cf1d9714d7
```
I think it would be nice to be consistent here, both by listing the full directory and using the same text.

File: src/node/blockstorage.cpp [hunk 2]
```diff
@@ -1156,9 +1157,41 @@ FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, cons
...
+static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
+{
+    // Bytes are serialized without length indicator, so this is also the exact
+    // size of the XOR-key file.
+    std::array<std::byte, 8> xor_key{};
+
+    // Initialize random fresh key, ...
+    FastRandomContext{}.fillrand(xor_key);
+    // ... but allow legacy or user-disabled key.
+    if (!fs::is_empty(opts.blocks_dir) || opts.xor_key == false) std::fill(xor_key.begin(), xor_key.end(), std::byte{});
```
Comment (TheCharlatan) [inline src/node/blockstorage.cpp:1169]: Nit: I found this a bit hard to follow and a bit unfortunate that a new random key would be generated every time whether it were used or not. How about:
```diff
@@ -1166,4 +1165,0 @@ static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
-    // Initialize random fresh key, ...
-    FastRandomContext{}.fillrand(xor_key);
-    // ... but allow legacy or user-disabled key.
-    if (!fs::is_empty(opts.blocks_dir) || opts.xor_key == false) std::fill(xor_key.begin(), xor_key.end(), std::byte{});
@@ -1171 +1167,8 @@ static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
-    if (fs::exists(xor_key_path)) {
+    bool key_exists = fs::exists(xor_key_path);
+
+    if (!key_exists && opts.xor_key && fs::is_empty(opts.blocks_dir)) {
+        // Initialize random fresh key, ...
+        FastRandomContext{}.fillrand(xor_key);
+    }
+
+    if (key_exists) {
```

**Review by sedited — APPROVED** [review]
ACK faaf3cf535999c2c9a84337414c8b35435384862

File: src/node/blockstorage.cpp [hunk 3]
```diff
@@ -1156,9 +1157,41 @@ FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, cons
...
+    const fs::path xor_key_path{opts.blocks_dir / "xor.dat"};
+    if (fs::exists(xor_key_path)) {
+        // A pre-existing xor key file has priority.
+        AutoFile xor_key_file{fsbridge::fopen(xor_key_path, "rb")};
+        xor_key_file >> xor_key;
```
[the remark this answers is not in the document]
Comment (hodlinator) [reply src/node/blockstorage.cpp:1180]: Tested `bitcoind -blocksxor=01234567`, which does not throw. Evaluates to true. Discussion seems to have stalled since #12713 https://github.com/bitcoin/bitcoin/blob/6f9db1ebcab4064065ccd787161bf2b87e03cc1f/src/common/args.cpp#L43-L63

**Review by hodlinator — COMMENTED** [review]
Concept ACK fae8e0a9825d107431f3f33e4f70fe54d02ab3e3
(Haven't had time to look at the tests yet).

Appreciate the robustness of `InitBlocksdirXorKey()` after some testing of removing the `blocks/xor.dat` file, re-running `bitcoind`. `hexdump`ing it, removing the whole `blocks/` directory, rerunning, `hexdump`ing again etc.

File: src/kernel/blockmanager_opts.h [hunk 4]
```diff
@@ -14,12 +14,15 @@ class CChainParams;
 
 namespace kernel {
 
+static constexpr bool DEFAULT_XOR_KEY_BLOCKSDIR{true};
```
[the remark this answers is not in the document]
Comment (hodlinator) [reply src/kernel/blockmanager_opts.h]: Good to see the removal of ` == true` as well in `InitBlocksdirXorKey()`. :+1:

File: test/functional/feature_reindex.py [hunk 5]
```diff
@@ -55,12 +65,15 @@ def find_block(b, start):
...
 
             # Swap the second and third blocks (don't disturb the genesis block).
+            b_new = bytearray(b)
+            b_new[b2_start:b3_start] = b[b3_start:b4_start]
+            b_new[b3_start:b4_start] = b[b2_start:b3_start]
+            b_new = util_xor(b_new, xor_dat)
             bf.seek(b2_start)
-            bf.write(b[b3_start:b4_start])
-            bf.write(b[b2_start:b3_start])
+            bf.write(b_new[b2_start:b4_start])
```
Comment (hodlinator) [inline test/functional/feature_reindex.py:76]: Smaller change if we make `util_xor()` take an offset:
```diff
diff --git a/test/functional/feature_reindex.py b/test/functional/feature_reindex.py
index f079a5e343..9f3f5357ba 100755
--- a/test/functional/feature_reindex.py
+++ b/test/functional/feature_reindex.py
@@ -43,15 +43,15 @@ class ReindexTest(BitcoinTestFramework):
             NUM_XOR_BYTES = 8 # From InitBlocksdirXorKey::xor_key.size()
             xor_dat = xor_f.read(NUM_XOR_BYTES)
 
-        def util_xor(data, key):
+        def util_xor(data, key, offset):
             data = bytearray(data)
             for i in range(len(data)):
-                data[i] ^= key[i % len(key)]
+                data[i] ^= key[(i + offset) % len(key)]
             return bytes(data)
 
         with open(blk0, 'r+b') as bf:
             # Read at least the first few blocks (including genesis)
-            b = util_xor(bf.read(2000), xor_dat)
+            b = util_xor(bf.read(2000), xor_dat, 0)
 
             # Find the offsets of blocks 2, 3, and 4 (the first 3 blocks beyond genesis)
             # by searching for the regtest marker bytes (see pchMessageStart).
@@ -68,12 +68,9 @@ class ReindexTest(BitcoinTestFramework):
             assert_equal(b3_start - b2_start, b4_start - b3_start)
 
             # Swap the second and third blocks (don't disturb the genesis block).
-            b_new = bytearray(b)
-            b_new[b2_start:b3_start] = b[b3_start:b4_start]
-            b_new[b3_start:b4_start] = b[b2_start:b3_start]
-            b_new = util_xor(b_new, xor_dat)
             bf.seek(b2_start)
-            bf.write(b_new[b2_start:b4_start])
+            bf.write(util_xor(b[b3_start:b4_start], xor_dat, b2_start))
+            bf.write(util_xor(b[b2_start:b3_start], xor_dat, b3_start))
 
         # The reindexing code should detect and accommodate out of order blocks.
         with self.nodes[0].assert_debug_log([
```

**Review by hodlinator — APPROVED** [review]
ACK fa986c918a239cabda198f044b6eb9739c298be9

**Review by hodlinator — APPROVED** [review]
re-ACK fa9d4aa9a1a8df6343f615572479db9c8e26b19d

**Review by sedited — APPROVED** [review]
Re-ACK fa51f311ee4f2295c7bd82f46e27dc043a33f955

File: src/node/blockstorage.cpp (hunk 2 above)
Comment (TheCharlatan) [reply src/node/blockstorage.cpp]: Thanks, this is clearer now.

**Review by sedited — APPROVED** [review]
Re-ACK fa895c72832f9555b52d5bb1dba1093f73de3136

Just a typo fix since last push.

**Review by hodlinator — APPROVED** [review]
ACK fa895c72832f9555b52d5bb1dba1093f73de3136

`git range-diff fa9d4aa9a1a8df6343f615572479db9c8e26b19d~3 fa9d4aa9a1a8df6343f615572479db9c8e26b19d fa895c72832f9555b52d5bb1dba1093f73de3136`

Confirmed only thing that really changed was:
* Merging with mainly 7aa8994c6fceae5cf8fb7e661371cdb19d2cb482 - " refactor: Add FlatFileSeq member variables in BlockManager "
* Correcting CAutoFile -> AutoFile in commit message
* Comment typo in `InitBlocksdirXorKey`
* `LogInfo` format string re-jiggling.

Passed `make check` & `test_runner.py`.

---

It would be nice to get to the bottom with why
```C++
LogInfo("Using obfuscation key for %s/*.dat files: '%s'\n", fs::PathToString(opts.blocks_dir), HexStr(xor_key));
```
was worse than
```C++
LogInfo("Using obfuscation key for blocksdir *.dat files (%s): '%s'\n", fs::PathToString(opts.blocks_dir), HexStr(xor_key));
```
and fix **test/lint/run-lint-format-strings.py** in the longer term.

**Comment by achow101** [comment]
ACK fa895c72832f9555b52d5bb1dba1093f73de3136

**Comment by hodlinator** [comment]
Is there some label or something that can assist people in discovering this kind of follow-up work on closed/merged PRs?

**Review by Scutua — APPROVED** [review]

**Review by Scutua — APPROVED** [review]

**Comment by Ronlabrie7470** [comment]
> Currently the *.dat files in the blocksdir store the data received from remote peers as-is. This may be problematic when a program other than Bitcoin Core tries to interpret them by accident. For example, an anti-virus program or other program may scan them and move them into quarantine, or delete them, or corrupt them. This may cause Bitcoin Core to fail a reorg, or fail to reply to block requests (via P2P, RPC, REST, ...).
> 
> Fix this, similar to https://github.com/bitcoin/bitcoin/pull/6650, by rolling a random XOR pattern over the dat files when writing or reading them.
> 
> Obviously this can only protect against programs that accidentally and unintentionally are trying to mess with the dat files. Any program that intentionally wants to mess with the dat files can still trivially do so.
> 
> The XOR pattern is only applied when the blocksdir is freshly created, and there is an option to disable it (on creation), so that people can disable it, if needed.

414bfe58b27d8bdc4ef1292bd0accd27cee0665f
